src/perl/lib/WTSI/NPG/Genotyping/SNPSet.pm

```diff
@@ -80,6 +82,77 @@ sub BUILD {
   }
 }
 
+
+=head2 union
+
+  Arg [1]    : ArrayRef[WTSI::NPG::Genotyping::SNPSet]
+  Arg [2]    : Maybe[Str]
+
+  Example    : my $union_set = $snpset->union(other_snpsets)
+  Description: Return a merged SNPSet object, containing the union of all
+               SNPs and
+               references in self and the ArrayRef argument. Optionally,
+               supply a name for the union SNPSet. Raise an error if column
+               names of input SNPSets are not consistent.
+  Returntype : WTSI::NPG::Genotyping::SNPSet
+
+=cut
+
+sub union {
+  my ($self, $others, $name) = @_;
+  my $colnames = Set::Scalar->new(@{$self->column_names});
+  # can't use Set::Scalar for Moose objects; use hashes instead
+  # Distinct SNPs may share a name (eg. gender markers)
+  # so, identify SNPs by chromosome and position
+  # if references are consistent, chromosome names should also be consistent
+  # if not, SNPs will be duplicated and SNPSet constructor will throw an error
+  my %union_snps;
+  my %union_refs;
+  foreach my $snp (@{$self->snps}) {
+      $union_snps{$snp->chromosome.':'.$snp->position} = $snp;
+  }
+  foreach my $ref (@{$self->references}) {
+      $union_refs{$ref->name} = $ref;
+  }
+  foreach my $other (@{$others}) {
+      my $other_colnames = Set::Scalar->new(@{$other->column_names});
+      unless ($colnames->is_equal($other_colnames)) {
+          $self->logcroak("Cannot merge SNPSet objects with ",
+                          "non-identical column names");
+      }
+      foreach my $snp (@{$other->snps}) {
+          my $snp_id = $snp->chromosome.':'.$snp->position;
+          if ($union_snps{$snp_id}) {
+              if (!$snp->equals($union_snps{$snp_id})) {
+                  $self->logcroak("Non-equal SNPs with same chromosome ",
+                                  "and position: '", $snp_id, "'");
+              }
+          } else {
+              $union_snps{$snp_id} = $snp;
+          }
+      }
+      foreach my $ref (@{$other->references}) {
+          $union_refs{$ref->name} = $ref;
+      }
+  }
+  my @colnames = $colnames->members;
+  my @snps = values %union_snps;
+  my @refs = values %union_refs;
+  # nasty hack to create a file as required by the Storable role
+  # TODO find a better solution
+  my $dir = tempdir( CLEANUP => 1 );
+  my $tempfile = $dir.'/snpset.tsv';
+  system("touch $tempfile");
```

Comment:
Prefer File::Temp to a system call as it provides safe options (this way is also unsafe because the exit code isn't checked).